### PR TITLE
feat(Notifications): new notifications FE plugin, API and backend

### PR DESCRIPTION
**Pull request by mareklibra**
Closes https://github.com/janus-idp/backstage-plugins/issues/817

This pull-request adds:
- BE plugin exposing new `REST API` for sending and querying notifications
- FE plugin
  - showing the notifications from the logged-in user's perspective (`NotificationsPage`)
  - FE `NotificationsApi` for accessing the notifications by 3rd-party FE plugins

Implemented as a static plugin since that contribution is planed to become part of the default Backstage deployment.

The scope of this PR is MVP, we already have a list of additions to be implemented in follow-ups.
Our recent intention is to provide the MVP to other projects, gather feedback and farther improve.

### CODE REVIEW

**Comment by mareklibra** [comment]
Screenshot from 2023-11-14 12-01-36

**Comment by pkliczewski** [comment]
@christophe-f fyi

File: plugins/notifications-backend/src/service/db.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+      table.text('message');
+      table.string('topic');
+      table.boolean('is_system').notNullable(); // is it a system message or a message for specific users and groups
+    });
+  }
+
+  if (!(await dbClient.schema.hasTable('users'))) {
+    await dbClient.schema.createTable('users', table => {
+      table.uuid('message_id').notNullable();
+      table.string('user').notNullable();
```
Comment (masayag) [inline plugins/notifications-backend/src/service/db.ts:36]: @ydayagi should there be an index created for the user column since it is being used to query?

**Comment by christophe-f** [comment]
> @christophe-f fyi

Thanks @pkliczewski

File: plugins/notifications-backend/src/service/db.ts (hunk 1 above)
Comment (ydayagi) [reply plugins/notifications-backend/src/service/db.ts:36]: https://issues.redhat.com/browse/FLPATH-727

**Comment by mareklibra** [comment]
Cc @kadel , @tumido May I ask for a review here, please?
It's difficult to keep rebasing the PR.

A little bit of context:
We have more feature requests coming and would prefer to phase their implementation.
There are two immediate projects waiting for the Notifications. In addition, we would like to kick-off the effort on upstream contribution based on that.

File: plugins/notifications-backend/src/service/handlers.ts [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+// createNotification
+// returns string id of created notification
+export async function createNotification(
+  dbClient: Knex<any, any>,
+  catalogClient: CatalogClient,
+  req: CreateNotificationRequest,
+): Promise<{ messageId: string }> {
+  let isUser = false;
+
+  // validate users
```
Comment (drodil) [inline plugins/notifications-backend/src/service/handlers.ts:31]: I would like to see that the users of target groups are just appended to the targetUsers and the groups are not even saved to the database. This allows to only target notifications to specific users and that way it's possible to mark the status for the notifications as in `new`, `read`, `discarded` etc.

File: plugins/notifications-backend/src/service/db.ts [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  const dbClient = knex(knexConfig);
+
+  // create tables
+  if (!(await dbClient.schema.hasTable('messages'))) {
+    await dbClient.schema.createTable('messages', table => {
+      table.uuid('id', { primaryKey: true }).defaultTo(dbClient.fn.uuid());
+      table.string('origin').notNullable();
+      table.timestamp('created').defaultTo(dbClient.fn.now()).index();
+      table.string('title').notNullable();
+      table.text('message');
```
Comment (drodil) [inline plugins/notifications-backend/src/service/db.ts:25]: I would like to see also `status` field for the messages. When the user fetches the messages, the system could automatically mark those as `read` and indicate that later in the UI which are new and which are already read messages. What do you think?

File: plugins/notifications-backend/src/service/db.ts (hunk 1 above)
Comment (ydayagi) [reply plugins/notifications-backend/src/service/db.ts]: fixed

File: plugins/notifications-backend/src/service/router-test-ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
+import { getVoidLogger } from '@backstage/backend-common';
```
Comment (gashcrumb) [inline plugins/notifications-backend/src/service/router-test-ts:1]: Is this file the right name?

Comment (mareklibra) [reply plugins/notifications-backend/src/service/router-test-ts:1]: nope, committed by mistake, will be removed

**Comment by mareklibra** [comment]
Rebased

File: plugins/notifications-backend/src/service/router-test-ts (hunk 4 above)
Comment (mareklibra) [reply plugins/notifications-backend/src/service/router-test-ts]: done

**Comment by christophe-f** [comment]
LGTM

**Review by christophe-f — APPROVED** [review]

**Comment by pkliczewski** [comment]
:tada:

**Pull request merged**